File: app/src/main/leveling.ts
import { type ChildProcess, spawn } from 'node:child_process'
import type {
  ApplyResult, AudioState, AutoResult, LevelEvent, Measurement, Paths, PresetFolder, PresetState,
  ReportResult, RiffList, SampleState, SceneRow
} from '../shared/types.js'
import { exists } from './util.js'

/**
 * Supervises the preset-leveling bench — `qc-mcp --leveling`, one more client
 * attached to the running daemon.
 *
 * It is a child process rather than TypeScript talking HID because everything
 * the bench needs already exists in Python: the framing, the protobuf pool, the
 * preset model and the parameter tapers. Re-implementing those here to move a
 * single dB value would be a second, diverging copy of the hardest code in the
 * repo.
 *
 * The wire is newline-delimited JSON, the same shape the daemon speaks: every
 * reply carries the `id` of its request, and anything with an `event` key is
 * unsolicited (meters, or a fatal that means the bench could not attach).
 */
export class Leveling {
  private child: ChildProcess | null = null
  private buf = ''
  private seq = 0
  private pending = new Map<number, { ok: (v: Reply) => void; fail: (e: Error) => void }>()
  private listeners = new Set<(e: LevelEvent) => void>()
  private stderr = ''

  constructor(private paths: Paths) {}

  setPaths(paths: Paths): void { this.paths = paths }

  get running(): boolean { return this.child !== null }

  onEvent(cb: (e: LevelEvent) => void): () => void {
    this.listeners.add(cb)
    return () => { this.listeners.delete(cb) }
  }

  private emit(e: LevelEvent): void {
    this.listeners.forEach((f) => f(e))
  }

  /** Idempotent: a second start on a live bench is a no-op, not a second child. */
  start(): void {
    if (this.child) return
    if (!exists(this.paths.bin)) {
      this.emit({ event: 'fatal', error: `qc-mcp is not installed at ${this.paths.bin}.` })
      return
    }
    this.stderr = ''
    const child = spawn(this.paths.bin, ['--leveling', '--socket', this.paths.socket], {
      cwd: this.paths.repo,
      stdio: ['pipe', 'pipe', 'pipe']
    })
    this.child = child
    child.stdout?.setEncoding('utf8')
    child.stdout?.on('data', (d: string) => this.feed(d))
    child.stderr?.on('data', (d) => { this.stderr += String(d) })
    child.on('error', (e) => { this.stderr += String(e) })
    child.on('exit', () => {
      this.child = null
      // Fail every in-flight call, or a renderer await hangs for ever.
      const why = this.stderr.trim().split('\n').slice(-2).join(' ').slice(0, 300)
      this.pending.forEach((p) => p.fail(new Error(why || 'the leveling bench exited')))
      this.pending.clear()
      this.emit({ event: 'stopped', error: why || null })
    })
  }

  stop(): void {
    const child = this.child
    this.child = null
    this.pending.forEach((p) => p.fail(new Error('the leveling bench was stopped')))
    this.pending.clear()
    if (child) {
      try { child.stdin?.end() } catch { /* already closed */ }
      setTimeout(() => { try { child.kill('SIGKILL') } catch { /* gone */ } }, 1500)
    }
  }

  private feed(chunk: string): void {
    this.buf += chunk
    let nl: number
    while ((nl = this.buf.indexOf('\n')) >= 0) {
      const line = this.buf.slice(0, nl).trim()
      this.buf = this.buf.slice(nl + 1)
      if (!line) continue
      let msg: Reply
      try { msg = JSON.parse(line) as Reply } catch { continue }
      if (typeof msg.event === 'string') { this.emit(msg as unknown as LevelEvent); continue }
      const waiter = this.pending.get(msg.id as number)
      if (!waiter) continue
      this.pending.delete(msg.id as number)
      waiter.ok(msg)
    }
  }

  /** One request/reply. Rejects on a bench-side error so the renderer can toast it. */
  call(op: string, args: Record<string, unknown> = {}, timeoutMs = 30000): Promise<Reply> {
    if (!this.child) this.start()
    const child = this.child
    if (!child) return Promise.reject(new Error('the leveling bench is not running'))
    const id = ++this.seq
    return new Promise<Reply>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(id)
        reject(new Error(`the bench did not answer '${op}' within ${Math.round(timeoutMs / 1000)}s`))
      }, timeoutMs)
      this.pending.set(id, {
        ok: (v) => {
          clearTimeout(timer)
          if (v.ok === false) reject(new Error(String(v.error ?? `'${op}' failed`)))
          else resolve(v)
        },
        fail: (e) => { clearTimeout(timer); reject(e) }
      })
      try {
        child.stdin?.write(`${JSON.stringify({ ...args, op, id })}\n`)
      } catch (e) {
        clearTimeout(timer)
        this.pending.delete(id)
        reject(e as Error)
      }
    })
  }

  // ── typed calls ───────────────────────────────────────────────────────
  //
  // The bench speaks the repo's snake_case; the renderer is given camelCase, so
  // the rename happens once, here, instead of in every component.

  async state(): Promise<PresetState> {
    return toPreset((await this.call('state')).preset)
  }

  async folders(refresh = false): Promise<PresetFolder[]> {
    const r = await this.call('folders', { refresh }, refresh ? 90000 : 30000)
    return (r.folders as RawFolder[]).map((f) => ({
      key: f.key,
      name: f.name,
      isFactory: f.is_factory,
      isDownloads: f.is_downloads,
      presets: f.presets.map((p) => ({
        position: p.position, name: p.name, cloudId: p.cloud_id
      }))
    }))
  }

  async open(
    folderKey: string, position: number, isFactory: boolean, cloudId: string
  ): Promise<PresetState> {
    const r = await this.call('open', {
      folder_key: folderKey, position, is_factory: isFactory, cloud_id: cloudId
    })
    return toPreset(r.preset)
  }

  async level(row: number, db: number): Promise<number> {
    return (await this.call('level', { row, db }, 8000)).db as number
  }

  async toggle(row: number, which: 'mute' | 'solo', on: boolean): Promise<boolean> {
    return (await this.call('switch', { row, which, on }, 8000)).on as boolean
  }

  async scene(index: number): Promise<number> {
    return (await this.call('scene', { index }, 8000)).scene as number
  }

  async save(name = ''): Promise<{ name: string; position: number }> {
    return (await this.call('save', { name })).saved as { name: string; position: number }
  }

  async meter(on: boolean): Promise<boolean> {
    return (await this.call('meter', { on }, 8000)).metering as boolean
  }

  // ── the measured half ───────────────────────────────────────────────────
  // These need the optional audio extra. The service answers with a plain
  // `available: false` rather than failing, so the view can explain itself.

  async audio(): Promise<AudioState> {
    const r = await this.call('audio', {}, 10000)
    return {
      available: r.available === true,
      error: r.error as string | undefined,
      hint: r.hint as string | undefined,
      quadCortex: (r.quad_cortex ?? null) as AudioState['quadCortex'],
      sample: (r.sample ?? null) as string | null
    }
  }

  async sampleArm(opts: { thresholdDbfs?: number; maxSeconds?: number } = {}) {
    return (await this.call('sample_arm', {
      threshold_dbfs: opts.thresholdDbfs ?? -40, max_seconds: opts.maxSeconds ?? 30
    }, 10000)) as unknown as SampleState
  }

  async sampleStatus(): Promise<SampleState> {
    return (await this.call('sample_status', {}, 8000)) as unknown as SampleState
  }

  /** Facts and envelope for the riff on disk, whoever recorded it and whenever. */
  async sampleInfo(): Promise<SampleState> {
    return (await this.call('sample_info', {}, 20000)) as unknown as SampleState
  }

  async sampleStop(): Promise<SampleState> {
    return (await this.call('sample_stop', {}, 20000)) as unknown as SampleState
  }

  async sampleDiscard(): Promise<SampleState> {
    return (await this.call('sample_discard', {}, 8000)) as unknown as SampleState
  }

  /**
   * Play the riff through the preset and measure what comes back. Writes nothing.
   *
   * No row: the service finds the lane the signal enters on and the one it leaves
   * by, which are routinely different. Naming a row here would only get that wrong.
   */
  async measure(perceived = false): Promise<Measurement> {
    const r = await this.call('measure', { perceived }, 120000)
    return (r.measurement ?? r) as unknown as Measurement
  }

  /**
   * Measure, trim, verify — until the preset lands on target.
   *
   * Slow by nature: every iteration plays the whole riff. The service emits an
   * `autolevel` event per pass so the view can follow rather than freeze, which
   * is why the timeout here is generous.
   */
  /** Play the riff through the preset so it can be heard. Returns once started. */
  async samplePlay(): Promise<{ playing: boolean; seconds: number }> {
    return (await this.call('sample_play', {}, 20000)) as unknown as
      { playing: boolean; seconds: number }
  }

  async sampleStopPlay(): Promise<void> {
    await this.call('sample_stop_play', {}, 8000)
  }

  /** Put back every fader an applied trim moved. */
  async revertLevels(): Promise<{ reverted: { row: number; db: number }[] }> {
    return (await this.call('revert_levels', {}, 15000)) as unknown as
      { reverted: { row: number; db: number }[] }
  }

  /** Measure every preset given and report what each needs. Writes nothing. */
  async measureMany(
    presets: { folder_key: string; position: number; name: string; cloud_id?: string }[],
    o: { target?: number; metric?: string } = {}
  ): Promise<ReportResult> {
    return (await this.call('measure_many', {
      presets, target: o.target ?? -18, metric: o.metric ?? 'lufs'
      // Generous: every preset is a recall plus a full playback of the riff.
    }, 600000)) as unknown as ReportResult
  }

  /** Measure every scene of the loaded preset. Writes nothing. */
  async measureScenes(o: { target?: number; scenes?: number[] } = {}):
    Promise<{ rows: SceneRow[] }> {
    return (await this.call('measure_scenes', {
      target: o.target ?? -18, scenes: o.scenes
    }, 600000)) as unknown as { rows: SceneRow[] }
  }

  /** Write the per-scene trims. Explicit: nothing here happens by default. */
  async levelScenes(o: { target?: number; scenes?: number[] } = {}): Promise<unknown> {
    return await this.call('level_scenes', {
      target: o.target ?? -18, scenes: o.scenes, dry_run: false
    }, 600000)
  }

  /**
   * Write one proposed correction, relative to where the fader is now.
   *
   * Separate from autolevel on purpose: the report shows a number, the user may
   * change it, and Apply has to write THAT. Re-measuring here would let the
   * table describe a change nobody is going to make.
   */
  /** The riffs the package ships, so nobody needs a guitar to start. */
  async riffs(): Promise<RiffList> {
    return (await this.call('riffs', {}, 30000)) as unknown as RiffList
  }

  async useRiff(name: string): Promise<SampleState> {
    // Rendering the first time takes a moment of pure numpy; after that it is
    // a file copy.
    return (await this.call('use_riff', { name }, 60000)) as unknown as SampleState
  }

  async applyTrim(o: {
    folderKey?: string; position?: number; isFactory?: boolean; cloudId?: string
    row?: number; db: number
  }): Promise<ApplyResult> {
    return (await this.call('apply_trim', {
      folder_key: o.folderKey, position: o.position,
      is_factory: o.isFactory ?? false, cloud_id: o.cloudId ?? '',
      row: o.row, db: o.db
    }, 120000)) as unknown as ApplyResult
  }

  async autolevel(opts: {
    target?: number; metric?: string; tolerance?: number; dryRun?: boolean
  } = {}): Promise<AutoResult> {
    return (await this.call('autolevel', {
      target: opts.target ?? -18, metric: opts.metric ?? 'lufs',
      // Reports by default: moving the player's faders is a separate decision
      // from measuring, so it has to be asked for.
      tolerance: opts.tolerance ?? 0.5, dry_run: opts.dryRun ?? true
    }, 300000)) as unknown as AutoResult
  }
}

interface RawLane {
  row: number; db: number; pan: number; mute: boolean; solo: boolean
  out_portid: number; out: string; active: boolean; physical: boolean; blocks: number
}

interface RawFolder {
  key: string; name: string; is_factory: boolean; is_downloads: boolean
  presets: { position: number; name: string; cloud_id: string }[]
}

function toPreset(raw: unknown): PresetState {
  const p = raw as {
    name: string; folder_key: string; position: number | null; is_factory: boolean
    scene: number; scene_labels: string[]; lanes: RawLane[]
  }
  return {
    name: p.name,
    folderKey: p.folder_key,
    position: p.position,
    isFactory: p.is_factory,
    scene: p.scene,
    sceneLabels: p.scene_labels ?? [],
    lanes: (p.lanes ?? []).map((l) => ({
      row: l.row, db: l.db, pan: l.pan, mute: l.mute, solo: l.solo,
      outPortId: l.out_portid, out: l.out, active: l.active,
      physical: l.physical, blocks: l.blocks
    }))
  }
}

export interface Reply {
  id?: number
  ok?: boolean
  error?: string
  event?: string
  [k: string]: unknown
}

File: src/qc_mcp/leveling.py
"""The preset-leveling service — the device half of Patchbay's Leveling bench.

Why this exists
---------------
Balancing a setlist means auditioning presets against each other and trimming
each one until they sit at the same loudness. On the device that is a lot of
walking: recall a preset, find its output block, nudge, recall the next one,
walk back. The bench collapses it into one screen — and this module is what
that screen talks to.

It is deliberately a *separate attached client*, not new MCP tools. The daemon
already fans device->host reports out to every client with a disjoint
`request_id` range, so the bench rides the same session Claude and Cortex
Control are on, and the user can keep editing in all three at once.

What "level" means here
-----------------------
The **Lane Output Control** (#23000, `Chain.output_control`) VOLUME parameter —
one per grid row. That is the value the leveling workflow wants because it is
stored *in the preset*, so trimming it and saving makes the balance permanent.
It is normalized 0..1 on the wire over a calibrated -40..+12 dB range
(`catalog.SYMBOLIC`), so the bench talks pure dB and converts at this edge.

Wire protocol — newline-delimited JSON on stdio, same shape as the daemon's.

    -> {"id": 1, "op": "hello"}
    <- {"id": 1, "ok": true, "firmware": "4.1.0", "mixer_db": [-40.0, 12.0]}
    -> {"id": 2, "op": "open", "folder_key": "...", "position": 3}
    <- {"id": 2, "ok": true, "preset": {...}}
    <- {"event": "meter", "outputs": {...}}          (unsolicited, while metering)

Threading: stdin is drained by one reader thread into a queue and **every**
device call happens on the main loop, because `QuadCortex._pending` is shared
state — a meter pump collecting on a second thread would eat the replies a read
is waiting for. See `transport._serialized` for the same trap one layer down.
"""
from __future__ import annotations

import json
import queue
import sys
import threading
import time

from . import catalog, directory
from . import protocol as P

#: LaneOutputControl parameter order.
VOLUME, PAN, MUTE, SOLO = 0, 1, 2, 3
LANE_OUTPUT_HASH = 23000

#: How often the loop pumps the device for streamed telemetry, in seconds.
PUMP = 0.04

_IOMETER = P.NAME_TO_CMD["IOMeter"]

#: Physical destinations a lane can be routed to, for the bench's lane labels.
#: `out_portid` values observed on a QC; unknown ids fall back to "Out <id>".
# NB the em-dash label is fine: it travels as JSON to the renderer and is drawn
# in HTML, never printed to a Windows console (the ASCII rule is about consoles).
OUT_PORTS = {
    0: "—", 1: "Out 1/2", 2: "Out 3/4", 3: "Send 1/2",
    16: "Mix bus", 19: "Multi Out",
}

#: Destinations that are NOT a way out of the preset: 0 is unrouted, 16 merges
#: into another row. Levelling a preset means moving what feeds the jacks, so the
#: bench's master knob skips these — moving a merge bus as well would apply the
#: same trim twice to any signal that passes through it.
INTERNAL_OUTS = {0, 16}

#: The IOMeter fields the bench shows, in the order it shows them.
METER_FIELDS = ("xlr_1", "xlr_2", "out_3", "out_4", "hp_l", "hp_r")
LIMITER_FIELDS = {"xlr_1": "xlr_1_limiter", "xlr_2": "xlr_2_limiter",
                  "out_3": "out_3_limiter", "out_4": "out_4_limiter"}


def db_of(norm):
    """Stored 0..1 -> dB on the lane output's calibrated range."""
    return catalog.to_display(LANE_OUTPUT_HASH, VOLUME, float(norm))


def norm_of(db):
    """dB -> the stored 0..1 the device wants."""
    return catalog.to_norm(LANE_OUTPUT_HASH, VOLUME, float(db))


def db_range():
    lo = catalog.SYMBOLIC["MIN_MIXER_DB"]
    hi = catalog.SYMBOLIC["MAX_MIXER_DB"]
    return lo, hi


# ─────────────────────────────────────────────────────────── reading state ────

def _lane(pos, chain, scene=0):
    """One grid row's output block as the bench sees it.

    NB the read-vs-delta indexing rule: in a whole-preset read the array
    *position* is the row and `chain.row` is 0, so the caller passes `pos`.

    `param_values` holds one entry per scene. Reading [0] unconditionally shows
    scene A's level whatever scene the device is actually on, which is wrong for
    any preset that scenes its output level.
    """
    ocs = list(chain.output_control)
    if not ocs:
        return None
    params = list(ocs[0].params)

    def value(idx, default=0.0):
        if idx >= len(params) or not params[idx].param_values:
            return default
        vals = params[idx].param_values
        return float(vals[scene if scene < len(vals) else 0].float_value)

    out_id = int(getattr(chain, "out_portid", 0) or 0)
    return {
        "row": pos,
        "db": round(db_of(value(VOLUME, norm_of(0.0))), 2),
        # PAN is stored 0..1 around centre; the app shows it as 50L..50R.
        "pan": round((value(PAN, 0.5) - 0.5) * 100.0),
        "mute": value(MUTE) >= 0.5,
        "solo": value(SOLO) >= 0.5,
        "out_portid": out_id,
        "out": OUT_PORTS.get(out_id, f"Out {out_id}"),
        # A lane routed nowhere is still in the preset but makes no sound, so the
        # bench shows it greyed rather than hiding it (hiding it would renumber
        # the rows and make the row indices lie).
        "active": out_id != 0,
        # …and only a lane that reaches an output counts toward the preset's level
        "physical": out_id not in INTERNAL_OUTS,
        "blocks": sum(1 for m in chain.models if m.hash),
    }


class Bench:
    """One attached session, plus the little bit of state the bench needs."""

    def __init__(self, qc):
        self.qc = qc
        self.metering = False
        self._last_meter = 0.0

    # -- device reads ------------------------------------------------------

    def preset_state(self):
        """Everything one bench column shows about the loaded preset.

        A whole-preset read can come back empty when the session is busy (two
        clients reading at once will do it), so retry through a reconnect once
        and then say so plainly — an `AttributeError` on `None.chains` tells the
        user nothing.
        """
        bp = self.qc.get_current_preset()
        if bp is None:
            try:
                self.qc.reconnect()
            except Exception:
                pass
            bp = self.qc.get_current_preset()
        if bp is None:
            raise RuntimeError(
                "the device did not answer a preset read; it may be busy - try again")
        pos_info = self.qc.get_setlist_position() or {}
        return self._state(bp, folder_key=pos_info.get("folder_key", ""),
                           position=pos_info.get("position"),
                           is_factory=pos_info.get("is_factory", False))

    def _state(self, bp, folder_key, position, is_factory):
        """One bench column's view of a preset. `bp` is a BinaryPreset, however
        it was obtained — a read, or the device's own recall broadcast."""
        scene = self.current_scene()
        lanes = []
        for pos, ch in enumerate(bp.chains):
            lane = _lane(pos, ch, scene)
            if lane:
                lanes.append(lane)
        return {
            "name": getattr(bp, "name", "") or "",
            "folder_key": folder_key or "",
            "position": position,
            "is_factory": bool(is_factory),
            "scene": scene,
            "scene_labels": [s for s in getattr(bp, "scene_labels", [])],
            "lanes": lanes,
        }

    def current_scene(self, default=0):
        """The active scene, best-effort.

        A Scene READ is not always answered promptly — and a preset's level is
        still perfectly editable when we don't know which scene is up — so this
        never fails the whole read the way an uncaught `QCError` would.
        """
        try:
            scene = self.qc.read_state("Scene", timeout_ms=1500)
        except Exception:
            return default
        return int(getattr(scene, "selected_scene", default)) if scene else default

    def folders(self, refresh=False):
        """Preset folders and their occupied slots, for the bench's picker.

        Prefers the on-disk snapshot because a live listing costs ~12s of
        streamed `File` messages; falls back to the live read when there is no
        snapshot yet, and keeps the result for next time.
        """
        cat = None if refresh else _snapshot()
        if not cat or not cat.get("presets"):
            try:
                live = self.qc.list_directory()
            except Exception:
                live = None
            if live and live.get("presets"):
                _save_snapshot(live)
                cat = live
        if not cat:
            cat = {"presets": []}
        out = []
        for folder in cat.get("presets", []):
            files = [{"position": f["index"], "name": f["name"],
                      "cloud_id": f.get("cloud_id", "") or ""}
                     for f in folder.get("files", []) if f.get("name")]
            if not files:
                continue
            out.append({
                "key": folder["key"],
                "name": folder.get("name") or folder["key"],
                "is_factory": bool(folder.get("is_factory")),
                # Downloads are addressed by cloud_id, not folder+position — the
                # bench has to carry that through or the recall is silently refused.
                "is_downloads": bool(folder.get("is_downloads")),
                "presets": sorted(files, key=lambda f: f["position"]),
            })
        return out

    # -- device writes -----------------------------------------------------

    def open(self, folder_key, position, is_factory=False, cloud_id="", settle=6.0):
        """Recall a preset, and build its state from the device's own answer.

        A recall makes the QC push the whole preset back — `RecallPreset` with a
        payload — so the bench LISTENS for that rather than asking. Waiting by
        polling the setlist pointer every 150ms and then re-reading the grid cost
        roughly thirty round trips for one preset change, every one of them
        landing on a device that was in the middle of loading; it was slow, and
        it made the device behave oddly while it worked.

        Falls back to a plain read if the broadcast never comes.
        """
        if cloud_id:
            self.qc.recall(downloads_key=cloud_id)
        else:
            self.qc.recall(folder_key=folder_key, position=int(position),
                           is_factory=bool(is_factory))

        want = P.NAME_TO_CMD["RecallPreset"]
        bp = None
        deadline = time.time() + settle
        while bp is None and time.time() < deadline:
            try:
                self.qc._collect(0.1)
            except Exception:
                break
            batch, self.qc._pending = self.qc._pending, []
            for cmd, obj, _raw, _pb in batch:
                if cmd == want and obj is not None and obj.HasField("preset"):
                    bp = obj.preset

        if bp is None:
            return self.preset_state()
        # The broadcast IS the confirmation the recall landed, so the pointer we
        # asked for is the pointer the device is on — no need to read it back.
        return self._state(bp, folder_key=folder_key,
                           position=None if cloud_id else int(position),
                           is_factory=bool(is_factory))

    def set_db(self, row, db):
        lo, hi = db_range()
        db = max(lo, min(hi, float(db)))
        self.qc.set_lane_param(int(row), "output_control", VOLUME, norm_of(db))
        return round(db, 2)

    def set_switch(self, row, which, on):
        idx = MUTE if which == "mute" else SOLO
        self.qc.set_lane_param(int(row), "output_control", idx, 1.0 if on else 0.0)
        return bool(on)

    def set_scene(self, index):
        self.qc.set_scene(int(index))
        # A scene switch must be *confirmed* before anything is written against
        # it; the bench's next level write would otherwise land on the old scene.
        self.qc._await_scene(int(index), timeout_s=2.0)
        return int(index)

    def save(self, name=""):
        pos = self.qc.get_setlist_position() or {}
        folder = pos.get("folder_key") or "/media/p4/Presets/My Presets"
        idx = int(pos.get("position", 0))
        if not name:
            bp = self.qc.get_current_preset()
            name = getattr(bp, "name", "") or ""
        if not name:
            raise ValueError("this slot has no preset name; pass one to save it")
        # File CREATE, never RecallPreset SAVE: the latter hangs the device on an
        # empty slot and needs a reboot.
        self.qc.write_preset_file(folder, idx, name)
        return {"name": name, "position": idx, "folder_key": folder}

    # -- streamed telemetry -------------------------------------------------

    def pump(self, emit):
        """Collect streamed frames and emit the newest meter reading.

        The device broadcasts telemetry with `request_id=0`, so the newest frame
        wins — never the first one buffered. Levels are **linear amplitude**
        (0..1), not dB; the bench passes them through and the UI converts, so
        this stays the one place that knows the wire format.

        Frames are not guaranteed: one bridge-mode session produced none at all
        (see PROTOCOL.md 5). Callers should render "no reading" as a resting
        state rather than treating it as a failure.
        """
        try:
            self.qc._collect(PUMP)
        except Exception:
            return
        # DRAIN, don't filter. The loop is single-threaded and only pumps
        # between commands, so nothing here is a reply anyone is waiting for.
        # Keeping "everything that isn't known telemetry" instead leaks: any
        # other broadcast the device makes piles up for ever, the list is
        # rescanned every tick, and the process grows in both memory and CPU
        # until it pegs a core.
        batch, self.qc._pending = self.qc._pending, []
        newest = None
        for cmd, obj, _raw, _pb in batch:
            if cmd == _IOMETER and obj is not None:
                newest = obj          # request_id=0 broadcast: the last one wins
        if newest is None:
            return
        outs = {}
        for f in METER_FIELDS:
            if not hasattr(newest, f):
                continue
            entry = {"level": float(getattr(newest, f))}
            lim = LIMITER_FIELDS.get(f)
            if lim and hasattr(newest, lim):
                entry["limit"] = float(getattr(newest, lim))
            outs[f] = entry
        self._last_meter = time.time()
        emit({"event": "meter", "at": self._last_meter, "outputs": outs})


def _snapshot_path():
    import os
    return os.environ.get("QC_CATALOG_JSON") or os.path.join(
        os.path.dirname(__file__), "..", "..", "interceptor", "catalog.json")


def _snapshot():
    """The gitignored on-disk DIRECTORY snapshot, if one has been written."""
    try:
        with open(_snapshot_path()) as fh:
            return json.load(fh)
    except (OSError, ValueError):
        return None


def _save_snapshot(cat):
    """Keep a good live listing warm — it holds personal library names, so it
    stays gitignored, exactly like the MCP server's own snapshot."""
    try:
        out = dict(cat)
        out.setdefault("_counts", directory.counts(cat))
        with open(_snapshot_path(), "w") as fh:
            json.dump(out, fh, indent=1)
    except (OSError, TypeError):
        pass


# ──────────────────────────────────────────────────────────────── the loop ────

def _stdin_reader(q):
    for line in sys.stdin:
        q.put(line)
    q.put(None)


def serve(socket_path):
    """Attach to the daemon and serve the bench on stdio until stdin closes."""
    from .daemon import attach

    out_lock = threading.Lock()

    def emit(obj):
        with out_lock:
            sys.stdout.write(json.dumps(obj) + "\n")
            sys.stdout.flush()

    try:
        qc = attach(socket_path)
    except Exception as exc:
        emit({"event": "fatal", "error": f"{type(exc).__name__}: {exc}"})
        return 1

    bench = Bench(qc)
    q: queue.Queue = queue.Queue()
    threading.Thread(target=_stdin_reader, args=(q,), daemon=True).start()

    ops = {
        "hello": lambda m: {"firmware": qc.firmware, "mixer_db": list(db_range())},
        "state": lambda m: {"preset": bench.preset_state()},
        "folders": lambda m: {"folders": bench.folders(bool(m.get("refresh")))},
        "open": lambda m: {"preset": bench.open(m["folder_key"], m["position"],
                                                m.get("is_factory", False),
                                                m.get("cloud_id", ""))},
        "level": lambda m: {"db": bench.set_db(m["row"], m["db"])},
        "switch": lambda m: {"on": bench.set_switch(m["row"], m["which"], m["on"])},
        "scene": lambda m: {"scene": bench.set_scene(m["index"])},
        "save": lambda m: {"saved": bench.save(m.get("name", ""))},
        "meter": lambda m: {"metering": _set_metering(bench, qc, m.get("on", True))},
    }

    while True:
        # Block for a command, and pump once per idle tick. `get_nowait` here
        # burns a core: the QC streams metronome ticks continuously, so the
        # transport almost always has something buffered and returns instantly
        # without its own sleep — the loop then spins as fast as it can and
        # hammers the daemon socket with reads.
        try:
            line = q.get(timeout=PUMP)
        except queue.Empty:
            bench.pump(emit)
            continue
        if line is None:
            break
        line = line.strip()
        if not line:
            continue
        try:
            msg = json.loads(line)
        except ValueError:
            continue
        rid = msg.get("id")
        op = msg.get("op")
        try:
            fn = ops.get(op)
            if fn is None:
                emit({"id": rid, "ok": False, "error": f"unknown op {op!r}"})
                continue
            emit({"id": rid, "ok": True, **fn(msg)})
        except Exception as exc:           # never let one bad call kill the bench
            emit({"id": rid, "ok": False, "error": f"{type(exc).__name__}: {exc}"})

    try:
        qc.close()
    except Exception:
        pass
    return 0


def _set_metering(bench, qc, on):
    """Ask the device to stream IOMeter (what Cortex Control sends on connect).

    Observed: the QC answers this with frames only while audio is actually
    moving — a silent rig produces none at all, on our subscription or the app's.
    """
    m = P.message_class("IOMeter")()
    m.action = P.ACTION["CREATE" if on else "DELETE"]
    m.request_id = qc.next_request_id()
    qc.send("IOMeter", m)
    bench.metering = bool(on)
    return bench.metering
